src/objective-c/GRPCClient/GRPCInterceptor.m

```diff
@@ -19,146 +19,276 @@
   id<GRPCInterceptorInterface> _nextInterceptor;
   id<GRPCResponseHandler> _previousInterceptor;
+  GRPCInterceptor *_thisInterceptor;
+  dispatch_queue_t _dispatchQueue;
+  NSArray<id<GRPCInterceptorFactory>> *_factories;
+  GRPCTransportId _transportId;
+  BOOL _shutDown;
 }
 
-- (instancetype)initWithNextInterceptor:(id<GRPCInterceptorInterface>)nextInterceptor {
+- (instancetype)initWithFactories:(NSArray<id<GRPCInterceptorFactory>> *)factories
+              previousInterceptor:(id<GRPCResponseHandler>)previousInterceptor
+                      transportId:(nonnull GRPCTransportId)transportId {
   if ((self = [super init])) {
-    _nextInterceptor = nextInterceptor;
+    if (factories.count == 0) {
+      [NSException raise:NSInternalInconsistencyException
+                  format:@"Interceptor manager must have factories"];
+    }
+    _thisInterceptor = [factories[0] createInterceptorWithManager:self];
+    if (_thisInterceptor == nil) {
+      return nil;
+    }
+    _previousInterceptor = previousInterceptor;
+    _factories = factories;
+    // Generate interceptor
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000 || __MAC_OS_X_VERSION_MAX_ALLOWED >= 101300
+    if (@available(iOS 8.0, macOS 10.10, *)) {
+      _dispatchQueue = dispatch_queue_create(
+          NULL,
+          dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_DEFAULT, 0));
+    } else {
+#else
+    {
+#endif
+      _dispatchQueue = dispatch_queue_create(NULL, DISPATCH_QUEUE_SERIAL);
+    }
+    dispatch_set_target_queue(_dispatchQueue, _thisInterceptor.dispatchQueue);
+    _transportId = transportId;
   }
-
   return self;
 }
 
-- (void)setPreviousInterceptor:(id<GRPCResponseHandler>)previousInterceptor {
-  _previousInterceptor = previousInterceptor;
-}
-
 - (void)shutDown {
   _nextInterceptor = nil;
   _previousInterceptor = nil;
+  _thisInterceptor = nil;
+  _shutDown = YES;
+}
+
+- (void)createNextInterceptor {
+  NSAssert(_nextInterceptor == nil, @"Starting the next interceptor more than once");
+  NSAssert(_factories.count > 0, @"Interceptor manager of transport cannot start next interceptor");
+  if (_nextInterceptor != nil) {
+    NSLog(@"Starting the next interceptor more than once");
+    return;
+  }
+  NSMutableArray<id<GRPCInterceptorFactory>> *interceptorFactories = [NSMutableArray
+      arrayWithArray:[_factories subarrayWithRange:NSMakeRange(1, _factories.count - 1)]];
+  while (_nextInterceptor == nil) {
+    if (interceptorFactories.count == 0) {
+      _nextInterceptor =
+          [[GRPCTransportManager alloc] initWithTransportId:_transportId previousInterceptor:self];
+      break;
+    } else {
+      _nextInterceptor = [[GRPCInterceptorManager alloc] initWithFactories:interceptorFactories
+                                                       previousInterceptor:self
+                                                               transportId:_transportId];
+      if (_nextInterceptor == nil) {
```

Comment:
Why would `_nextInterceptor` be nil here (other than out-of-memory case)? Should we assert that `_nextInterceptor != nil`?

Reply:
It is `nil` if when creating the interceptor from the factory, the factory returned `nil` interceptor, in case the factory just decides it's not appropriate to insert an interceptor in this call.